engine/src/player.ts

```diff
@@ -52,9 +52,48 @@ interface FederationCache {
   custom: boolean;
 }
 
+export enum AutoChargeMode {
+  autoChargeUpToAmount,
+  alwaysAsk,
+  declineIfNotFree,
+}
+
+export class AutoCharge {
+  static alwaysAsk = "ask";
+  static declineIfNotFree = "decline-cost";
+  mode: AutoChargeMode;
+  amount: number;
+
+  constructor(value: string) {
+    if (value == AutoCharge.alwaysAsk) {
+      this.mode = AutoChargeMode.alwaysAsk;
+    } else if (value == "0" || value == AutoCharge.declineIfNotFree) {
+      //0 is legacy value
+      this.mode = AutoChargeMode.declineIfNotFree;
+      this.amount = 1;
+    } else {
+      this.mode = AutoChargeMode.autoChargeUpToAmount;
+      this.amount = Number(value);
+    }
+  }
+
+  value(): string {
+    switch (this.mode) {
+      case AutoChargeMode.alwaysAsk:
+        return AutoCharge.alwaysAsk;
+      case AutoChargeMode.declineIfNotFree:
+        return AutoCharge.declineIfNotFree;
+      default:
+        return String(this.amount);
+    }
+  }
+}
```

Comment:
This will fail when loading from JSON;

```
player.settings = data.settings ?? player.settings;
```

Because `data.settings` will just be a plain object at that point.

In general I see too many classes!

`Settings` can just be a plain object:

```ts
interface Settings {
  // ...
}

class Player {
  settings: Settings = {
     autoIncome: false,
     ...
  }
}
```

What you can do for the autoCharge part is either:

```ts
type AutoCharge = "ask" | "decline-cost" | 1 | 2 | 3 | 4 | 5

enum AutoCharge {
  AlwaysAsk = "ask",
  DeclineIfCost = "decline-cost",
  UpTo1 = 1,
  UpTo2 = 2, 
  ...
}
```

The first solution, even if no enum, is ok! Because when you try to do this:

```ts
// Typo: "aks" instead of "ask"
if (player.settings.autoCharge === "aks") {
}
```

or this:

```ts
// Error: 0 is a legacy value
if (player.settings.autoCharge === 0) {
}
```

Then Typescript will still catch it, and it will fail the CI job.

For the migration to remove old legacy values, you can do it in `fromJSON`:

```ts
player.settings = data.settings ?? player.settings;

->

player.settings = data.settings ?? player.settings;
// Legacy value
if (player.settings.autoChargePower as AutoCharge | 0 === 0) {
  player.settings.autoChargePower = "decline-cost";
}
```

Reply:
yes, that was a good idea :smile: